c/rce_validator.c

```diff
@@ -0,0 +1,261 @@
+  ptr->start_point = 0;
+  ptr->max_cache_size = cache_size;
+
+  cursor->offset = 0;
+  cursor->size = len;
+  cursor->data_source = ptr;
+
+  err = CKB_SUCCESS;
+exit:
+  return err;
+}
+
+static int make_data_cursor(uint8_t* buffer, uint32_t cache_size, size_t index,
+                            size_t source, mol2_cursor_t* cursor) {
+  return _make_cursor(buffer, cache_size, index, source, cursor,
+                      read_from_cell_data);
+}
+
+static uint32_t read_from_witness(uintptr_t* arg, uint8_t* ptr, uint32_t len,
+                                  uint32_t offset) {
+  int err;
+  uint64_t output_len = len;
+  err = ckb_checked_load_witness(ptr, &output_len, offset, arg[0], arg[1]);
+  if (err != 0) {
+    DEBUG("Error reading witness!");
+    ckb_exit(ERROR_EOF);
+  }
+  return output_len;
+}
+
+static int make_witness_cursor(uint8_t* buffer, uint32_t cache_size,
+                               size_t index, size_t source,
+                               mol2_cursor_t* cursor) {
+  return _make_cursor(buffer, cache_size, index, source, cursor,
+                      read_from_witness);
+}
+
+int main() {
+  uint8_t type_id[32];
+  int err = ckb_load_type_id_from_script_args(0, type_id);
+  if (err != CKB_SUCCESS) {
+    return err;
+  }
+  err = ckb_validate_type_id(type_id);
+  if (err != CKB_SUCCESS) {
+    return err;
+  }
+
+  uint8_t flags = 0;
+  uint64_t len = 1;
+  err = ckb_load_script(&flags, &len, 32);
+  if (err != CKB_SUCCESS) {
+    DEBUG("Cannot load current script!");
+    return err;
+  }
+  if (len > 1) {
+    DEBUG("Current script is too large!");
+    return ERROR_ARGUMENTS_LEN;
+  }
+  bool append_only = (flags & FLAG_APPEND_ONLY) != 0;
+
+  uint8_t input_hash[SMT_KEY_BYTES];
+  memset(input_hash, 0, SMT_KEY_BYTES);
+  bool has_input = false, input_is_rule = false;
+
+  if (_ckb_has_type_id_cell(0, 1) == 1) {
+    has_input = true;
+    uint8_t input_data_buffer[MOL2_DATA_SOURCE_LEN(CACHE_SIZE)];
+    mol2_cursor_t input_cell_data;
+    err = make_data_cursor(input_data_buffer, CACHE_SIZE, 0,
+                           CKB_SOURCE_GROUP_INPUT, &input_cell_data);
+    CHECK(err);
+
+    RCDataType rc_data = make_RCData(&input_cell_data);
+    uint32_t item_id = rc_data.t->item_id(&rc_data);
+    if (item_id == RCDataUnionRule) {
+      RCRuleType rule = rc_data.t->as_RCRule(&rc_data);
+      mol2_cursor_t smt_root = rule.t->smt_root(&rule);
+      uint32_t read = mol2_read_at(&smt_root, input_hash, SMT_KEY_BYTES);
+      CHECK2(read == SMT_KEY_BYTES, ERROR_INVALID_MOL_FORMAT);
+      input_is_rule = true;
+    } else if (item_id == RCDataUnionCellVec) {
+      input_is_rule = false;
+    } else {
+      return ERROR_INVALID_MOL_FORMAT;
+    }
+  }
+
+  uint8_t output_data_buffer[MOL2_DATA_SOURCE_LEN(CACHE_SIZE)];
+  mol2_cursor_t output_cell_data;
+  err = make_data_cursor(output_data_buffer, CACHE_SIZE, 0,
+                         CKB_SOURCE_GROUP_OUTPUT, &output_cell_data);
+  CHECK(err);
```

Comment:
It seems that we don't allow destroy RCE rule cell in this validator? Is this a intentional feature?